plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -20,61 +20,127 @@
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName == null) {
+                    // this might have been just created, reload cache
+                    Map<String, String> mapping = stream(listDatasets(projectId))
+                            // will throw on collisions to avoid ambiguity
+                            .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                    remoteDatasetNames.putAll(mapping);
+                }
+                remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName != null) {
+                    return remoteDatasetName;
+                }
+            }
+            catch (RuntimeException e) {
+                throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + firstNonNull(e.getMessage(), e), e);
+            }
+        }
+
+        return datasetName;
     }
 
-    TableInfo getTable(TableId tableId)
+    String toRemoteTableName(String projectId, String remoteDatasetName, String tableName)
     {
-        TableId bigQueryTableId = tableIds.get(tableId);
-        Table table = bigQuery.getTable(bigQueryTableId != null ? bigQueryTableId : tableId);
-        if (table != null) {
-            tableIds.putIfAbsent(tableId, table.getTableId());
-            datasetIds.putIfAbsent(toDatasetId(tableId), toDatasetId(table.getTableId()));
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(remoteDatasetName, "remoteDatasetName is null");
+        requireNonNull(tableName, "tableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(tableName), "Expected table name from internal metadata to be lowercase: %s", tableName);
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable Map<String, String> mapping = remoteTableNames.getIfPresent(remoteDatasetName);
+                if (mapping != null && !mapping.containsKey(tableName)) {
+                    // this might have been just created, reload cache
+                    mapping = null;
+                }
+                if (mapping == null) {
+                    mapping = stream(listTables(DatasetId.of(projectId, remoteDatasetName), TableDefinition.Type.TABLE, TableDefinition.Type.VIEW))
+                            // will throw on collisions to avoid ambiguity
+                            .collect(toImmutableMap(table -> table.getTableId().getTable().toLowerCase(ENGLISH), table -> table.getTableId().getTable()));
+                    remoteTableNames.put(remoteDatasetName, mapping);
+                }
+                String remoteTableName = mapping.get(tableName);
```

Comment:
If you want to keep whole mapping under single key then you can make this code more readable (I think) by using `LoadingCache`

Reply:
Given https://github.com/trinodb/trino/pull/6748/files#r568412619 I think this void.